facets/core/prop.ts

```diff
@@ -0,0 +1,81 @@
+import { FacetPropType, NullableTypeOfPropType, TypeOfFacetPropType } from './propType';
+
+type IsNeverNull<T> = Extract<T, null> extends never ? true : false;
+
+export type AvailableCascadeStrategyForPropType<PropType extends FacetPropType> =
+	// Any prop type can use the `overwrite` strategy
+	| 'overwrite'
+	// Only props that are objects can object-merge during cascade" {...a, ...b}
+	| (TypeOfFacetPropType<PropType> extends Record<string, any> ? 'merge' : never)
+	// Only props that are arrays can array-merge during cascasde: [...a, ...b]
+	| (TypeOfFacetPropType<PropType> extends any[] ? 'extend' : never);
+
+export type CascadeStrategy = 'overwrite' | 'merge' | 'extend';
+
+export type FacetPropOptions<
+	// The PropType of the prop.
+	PropType extends FacetPropType,
+	// The `rootValue` provided, which may be `null`.
+	// We provide this as a type parameter so that TypeScript will infer its narrowest -- most
+	// specific -- type and preserve that information in the inferred return type. This lets us
+	// determine statically whether a facet prop will ever be null and adjust its TypeScript
+	// definitions accordingly.
+	RootValue extends NullableTypeOfPropType<PropType>,
+	// When we provide a defaultValue to a facet, it may be `null`.
+	DefaultValue = NullableTypeOfPropType<PropType>,
+> = {
+	// The value of this prop at the "root" of PubPub -- in the absence of any other facet instances
+	// that override it, what value does this prop have?
+	rootValue: RootValue;
+	// The default value of this prop, if not explicitly provided to createFacet(). By contrast with
+	// `rootValue`, this value will actually be stored on a facet instance.
+	defaultValue?: DefaultValue;
+	// Explains how this prop should cascade from higher to lower scopes.
+	cascade?: AvailableCascadeStrategyForPropType<PropType>;
+	// A human-readable label for this prop
+	label?: string;
+};
+
+export type FacetProp<
+	PropType extends FacetPropType = FacetPropType,
+	RootValue extends NullableTypeOfPropType<PropType> = NullableTypeOfPropType<PropType>,
+> =
+	// Everything in the FacetPropOptions above...
+	Omit<FacetPropOptions<PropType, RootValue>, 'cascade'> & {
+		__facetProp: true;
+		// Plus the provided propType.
+		propType: PropType;
+		// Plus a non-nullable cascade strategy with a default value applied.
+		cascade: CascadeStrategy;
+	};
+
+export type RootValueOfFacetProp<Prop extends FacetProp> = Prop['rootValue'];
+export type TypeOfFacetProp<Prop extends FacetProp> = TypeOfFacetPropType<Prop['propType']>;
+export type NullableTypeOfFacetProp<Prop extends FacetProp> = NullableTypeOfPropType<
+	Prop['propType']
+>;
+
+export type CascadedTypeOfFacetProp<Prop extends FacetProp> =
+	// If the root value of the prop is non-null...
+	IsNeverNull<RootValueOfFacetProp<Prop>> extends true
+		? // ...then we'll never get null as a cascaded value for this prop
+		  TypeOfFacetProp<Prop>
+		: // Otherwise, we might.
+		  NullableTypeOfFacetProp<Prop>;
```

Comment:
Does this mean that intermediate scopes (i.e. collection) won't be able to nullify facet prop values?

Reply:
Yes — in facets-world, `null` is a clear pane of glass through which values from higher scopes can shine through (I promise I will document all of this). In cases where you don't want this to happen, I imagine there being an attribute on the specific facet instance (i.e. the `FacetBinding`?) that says "don't accept a cascade from higher scopes" or equivalently "make me the root for this cascade", which would accomplish the same thing. We could also invent another special signal value that acts like `initial` in CSS. Probably best to see which fits the exact use cases that arise.